File: 백준/Gold/9663. N－Queen/N－Queen.py
def main(n):
    cols = [False] * n
    # +
    rup = [False] * (n * 2)
    # -
    rdown = [False] * (n * 2)

    li = []
    cnt = 0

    def recur(r):
        nonlocal cnt
        if len(li) == n:
            cnt += 1
            return

        for c in range(n):
            if cols[c]:
                continue

            if rup[r + c]:
                continue

            if rdown[r - c + n]:
                continue

            cols[c] = True
            rup[r + c] = True
            rdown[r - c + n] = True
            li.append((r, c))
            recur(r + 1)
            li.pop()
            cols[c] = False
            rup[r + c] = False
            rdown[r - c + n] = False

    recur(0)
    return cnt


n = int(input())
print(main(n))

File: 백준/Gold/9663. N－Queen/N－Queen.js
const fs = require('fs');

const read = fs.readFileSync('dev/stdin').toString().trim()

const n = parseInt(read);

const main = () => {
    const cols = Array.from({length: n}).fill(false);
    const upright = {};
    const downright = {};
    let cnt = 0
    
    const recur = (r) => {
        if (r == n){
            cnt += 1
            return;
        }
        
        for(let c = 0; c < n; c++ ){
            if (cols[c]) continue;
            if (r - c in upright) continue;
            if (r + c in downright) continue;
            
            cols[c] = true
            upright[r - c] = true
            downright[r + c] = true
            recur(r + 1)
            cols[c] = false
            delete upright[r - c];
            delete downright[r + c];
        }
    }
    
    recur(0)
    return cnt;
}

console.log(main())